core/src/com/unciv/logic/civilization/CivConstructions.kt

```diff
@@ -0,0 +1,165 @@
+            freeStatBuildingsProvided[stat] = hashSetOf()
+        }
+    }
+    
+    fun clone(): CivConstructions {
+        val toReturn = CivConstructions()
+        toReturn.civInfo = civInfo
+        toReturn.freeBuildings.putAll(freeBuildings)
+        toReturn.freeStatBuildingsProvided.putAll(freeStatBuildingsProvided)
+        toReturn.freeSpecificBuildingsProvided.putAll(freeSpecificBuildingsProvided)
+        toReturn.boughtItemsWithIncreasingPrice.add(boughtItemsWithIncreasingPrice.clone())
+        return toReturn
+    }
+    
+    fun setTransients(civInfo: CivilizationInfo) {
+        this.civInfo = civInfo
+        
+        // civInfo.boughtConstructionsWithGloballyIncreasingPrice deprecated since 3.16.15, this is replacement code
+            if (civInfo.boughtConstructionsWithGloballyIncreasingPrice.isNotEmpty()) {
+                for (item in civInfo.boughtConstructionsWithGloballyIncreasingPrice) {
+                    boughtItemsWithIncreasingPrice.add(item.key, item.value)
+                }
+                civInfo.boughtConstructionsWithGloballyIncreasingPrice.clear()
+            }
+        //
+        
+        // Deprecated variables in civ.policies since 3.16.15, this is replacement code
+            if (civInfo.policies.specificBuildingsAdded.isNotEmpty()) {
+                for ((building, cities) in civInfo.policies.specificBuildingsAdded) {
+                    for (cityId in cities) {
+                        if (building !in freeSpecificBuildingsProvided)
+                            freeSpecificBuildingsProvided[building] = hashSetOf()
+                        freeSpecificBuildingsProvided[building]!!.add(cityId)
+
+                        if (cityId !in freeBuildings)
+                            freeBuildings[cityId] = hashSetOf()
+                        freeBuildings[cityId]!!.add(building)
+                    }
+                }
+                civInfo.policies.specificBuildingsAdded.clear()
+            }
+        
+            if (civInfo.policies.cultureBuildingsAdded.isNotEmpty()) {
+                for ((cityId, building) in civInfo.policies.cultureBuildingsAdded) {
+                    freeStatBuildingsProvided[Stat.Culture]!!.add(cityId)
+                    
+                    if (cityId !in freeBuildings)
+                        freeBuildings[cityId] = hashSetOf()
+                    freeBuildings[cityId]!!.add(building)
+                }
+                civInfo.policies.cultureBuildingsAdded.clear()
+            }
+        //
+    }
+    
+    fun startTurn() {
+        tryAddFreeBuildings()
+    }
+    
+    fun tryAddFreeBuildings() {
+        addFreeStatsBuildings()
+        addFreeSpecificBuildings()
+    }
+    
+    fun getFreeBuildings(cityId: String): HashSet<String> {
+        val toReturn = freeBuildings[cityId] ?: hashSetOf()
+        for (city in civInfo.cities) {
+            toReturn.addAll(city.cityConstructions.freeBuildingsProvidedFromThisCity[city.id] ?: hashSetOf())
```

Comment:
If this function was called frequently I'd quibble about the useless hashset creation for most cities, but since it's a rare one the effect is very minor